SYCL/ESIMD/api/functional/type_coverage.hpp

```diff
@@ -86,28 +91,62 @@ template <tested_types required> auto get_tested_types() {
   }
 }
 
+enum class tested_dims { all };
+
+// Factory method to retrieve pre-defined values_pack, to have the same
+// default dimensions over the tests
+template <tested_dims required> auto get_tested_dimensions() {
+  if constexpr (required == tested_dims::all) {
+    return values_pack<1, 8, 16, 32>();
+  } else {
+    static_assert(required != required, "Unexpected tested dimension");
+  }
+}
+
 // Run action for each of types given by named_type_pack instance
-template <template <typename, typename...> class action,
-          typename... actionArgsT, typename... types, typename... argsT>
-inline bool for_all_types(const named_type_pack<types...> &typeList,
-                          argsT &&... args) {
+template <template <typename, int, typename...> class Action, int N,
+          typename... ActionArgsT, typename... Types, typename... ArgsT>
+inline bool for_all_types(const named_type_pack<Types...> &type_list,
+                          ArgsT &&... args) {
   bool passed{true};
 
   // run action for each type from types... parameter pack
-  size_t typeNameIndex = 0;
-
-  int packExpansion[] = {
-      (passed &= action<types, actionArgsT...>{}(std::forward<argsT>(args)...,
-                                                 typeList.names[typeNameIndex]),
-       ++typeNameIndex,
-       0 // Dummy initialization value
-       )...};
+  size_t type_name_index = 0;
+
+  ((passed &= Action<Types, N, ActionArgsT...>{}(
+        std::forward<ArgsT>(args)..., type_list.names[type_name_index]),
+    ++type_name_index),
```

Comment:
```suggestion
  (passed &= Action<Types, N, ActionArgsT...>{}(
        std::forward<ArgsT>(args)..., type_list.names[type_name_index]),
    ++type_name_index,
```
Please correct me if I'm wrong, but wouldn't it give the same result due to strict sequencing of every expression delimited by comma operator?
And one more thing that we need to check.. I'm not able to map the expansion pack context to the list of the allowed ones: see "expansion loci" in https://en.cppreference.com/w/cpp/language/parameter_pack. Looks like we need to add a comment stating the expansion pack context explicitly, just to be aware of possible problems on code update.

Reply:
when I try to apply this comment I've obtained this compilation error:
```
SYCL\ESIMD\api\functional\ctors/../type_coverage.hpp:119:4: error: expected ')'
   ...);
   ^
SYCL\ESIMD\api\functional\ctors/../type_coverage.hpp:116:3: note: to match this '('
  (passed &= Action<Types, N, ActionArgsT...>{}(
  ^
SYCL\ESIMD\api\functional\ctors/../type_coverage.hpp:116:4: error: expression contains unexpanded parameter pack 'Types'
  (passed &= Action<Types, N, ActionArgsT...>{}(
   ^                ~~~~~
SYCL\ESIMD\api\functional\ctors\ctor_default.cpp:128:13: note: in instantiation of function template specialization 'esimd_test::api::functional::for_all_types_and_dims<test, initializer, char, unsigned char, signed char, short, unsigned
      short, int, unsigned int, long, unsigned long, float, double, long long, unsigned long long, 1, 8, 16, 32, cl::sycl::queue &>' requested here
  passed &= for_all_types_and_dims<test, initializer>(types, dims, queue);
            ^
1 warning and 2 errors generated.
```

I think that final simplification of this solution due to I tried many different simplifications of this function and only this version works correctly and without compiled without errors

Reply:
OK, I see now, that's an unary right fold expression available since C++17.
Therefore something like:
```
  const auto types = named_type_pack<foo, boo, bar>("foo", "boo", "bar");
  for_all_types<Action, 1>(types);
```
will result in following `named_type_pack`:
```
  named_type_pack::names[3] == {"foo", "boo", "bar"};
```
and in the following unfolded expression:
```
  size_t type_name_index = 0;

  ((passed &= Action<foo, 1>{}("foo"),
    type_name_index = 1),
    (
      (passed &= Action<boo, 1>{}("boo"),
       type_name_index = 2),
      (
        (passed &= Action<bar, 1>{}("bar"),
         type_name_index = 3),
      )
    )
  )
```

And therefore brackets are required due to lower precedence value of the increment operator.
Am I right?

Reply:
I would suggest the following comment then:
```
  // The unary right fold expression is used for parameter pack expansion.
  // Every expression with comma operator is strictly sequenced, so we can 
  // increment safely. And of course the fold expression would not be optimized
  // out due to side-effects.
  // Additional pair of brackets is required because of precedence of increment
  // operator relative to the comma operator.
  //
  // Note that there is actually no difference in left or right fold expression
  // for the comma operator, as it would give the same order of actions
  // execution and the same order of the type name index increment: both the
  // "(expr0, (exr1, expr2))" and "((expr0, expr1), expr2)" would give the same
  //  result as simple "expr0, expr1, expr2"
```
Thanks for the great idea!

Reply:
yes, order is right, You is right.

Reply:
The comments was replaced in this commit: 9ce01972c9d4b14c835faf126b6913a107449067